### PR TITLE
Check duplicate clauses

### PR DESCRIPTION
**Pull request by blapid**
Hi,

I saw your CppCon presentation on YouTube and I was very impressed, cool concept!
After some testing and hacking around, I saw that it's possible to define duplicate function names, for example: 

```
struct Drawable : decltype(dyno::requires(
  "draw"_s = dyno::function<void (dyno::T const&, std::ostream&)>,
  "draw"_s = dyno::function<void (dyno::T const&, std::ostream&)>
)) { };
```

Would compile.

Likewise, this:

```
template <>
auto const dyno::concept_map<Drawable, Square> = dyno::make_concept_map(
  "draw"_s = [](Square const& /*square*/, std::ostream& out) {
    out << "square" << std::endl;
  },
  "draw"_s = [](Square const& /*square*/, std::ostream& out) {
    out << "square" << std::endl;
  }
);
```

Would compile.

Imho this doesn't seem like a desirable outcome  :)
This patch is a quick hack to raise a `static_assert` when the same function is defined twice.

Let me know what you think, and merge if you like it :)

P.S. This raises another question: Would it be useful to allow duplicate names but with different signatures? Seems theoretically possible.

### CODE REVIEW

**Comment by ricejasonf** [comment]
`hana::unique` requires elements to be presorted as it only check adjacent elements in the sequence.

I'd suggest a pure type solution since the "names" are always stateless. (I think)

Boost.Hana `map` has a check like this, but it only checks with a special debug mode flag.
https://github.com/boostorg/hana/blob/master/include/boost/hana/map.hpp#L238

**Comment by ldionne** [comment]
Thanks for the PR. I'm currently at a C++ Committee meeting so I don't have much time, but I'll look at this after the meeting.

**Comment by blapid** [comment]
@ricejasonf you're correct, `hana::unique` is inadequate here. Names do seem to be stateless as they're specialization of `hana::string`s without any members. 
Does this look better to you?

**Comment by ricejasonf** [comment]
I'd provide more feedback, but leaving comments on lines of code is disabled for me for some reason.

Below your duplicates check, `hana::make_map` is being called which already does this if you define `BOOST_HANA_CONFIG_ENABLE_DEBUG_MODE`. The check is expensive so it only does it in this "debug mode".

I'm guessing that the map is getting merged with `hana::insert` which will drop the duplicates so this is more or less not failing because of the implementation of the map merging in Hana.

**Comment by blapid** [comment]
Are you referring to the changes in `concept_map.hpp`? I think you're right about the insert dropping duplicates, which indeed makes that part unneeded.

The changes in `concept.hpp` seem to be working though. The CI failure:
```
/home/travis/build/ldionne/dyno/include/dyno/concept.hpp:155:37: note: in
      instantiation of template class
      'dyno::concept<dyno::concept<dyno::MoveConstructible, dyno::Destructible>,
      dyno::Destructible, dyno::Storable>' requested here
constexpr dyno::concept<Clauses...> requires(Clauses ...) {
```
Is expected because "destruct" is defined twice, once in `poly.ctor.copy.cpp:18` and again in `poly.hpp:69` (from `poly.ctor.copy.cpp:19`).

**Comment by ricejasonf** [comment]
@blapid Then perhaps just remove the references to `hana::unique` and squash your commits to a single commit. I'd be concerned about the compile-time performance, but it could be moot working with really small sets.

About allowing multiple overloads with the same name, I think it is a good idea. I don't know if that is difficult in the rest of Dyno, but you could alter your check to look at the whole pair instead of just the name. idk just a thought

:coffee:

**Comment by blapid** [comment]
Done.
I agree, compile-time should not be seriously affected unless large amount of functions are used.

I've been messing around with the multiple overloads idea and got it to *kinda* work. The idea is to add some information to the function name that allows for differentiating between overloads. One way to do this is to create some kind of compile time *mangling* which theoretically could be done with [ctti](https://github.com/Manu343726/ctti). This *kinda* works because it breaks if you use generic lambdas in your interface. You're welcome to have a look [here](https://github.com/blapid/dyno/commit/88a3a4e9a0e33a9cf7d65df61f729a3a392da34d), note that `overview.cpp` will compile but `readme.cpp` won't (because of the generic lambda).

File: include/dyno/concept.hpp [hunk 1]
```diff
@@ -63,6 +63,21 @@ namespace detail {
...
   };
+
+  template <typename... Clauses>
+  struct check_duplicate_names {
+    template <typename... Pairs>
+    struct check_duplicate_names_impl{};
+  
+    template <typename... Pairs>
+    struct check_duplicate_names_impl<const boost::hana::basic_tuple<Pairs...>> {
+      static constexpr bool value = boost::hana::detail::has_duplicates<
```
Comment (ldionne) [inline include/dyno/concept.hpp]: You know you shouldn't be using stuff from a library's `detail::` namespace, right? :-)

**Comment by ldionne** [comment]
Regarding why it works right now, the "documentation" for `complete_concept_map` says:

> // [...]
// Furthermore, if the same function is defined in more than one concept map
// in the full refinement tree, it is undefined which one is used. Therefore,
// all of the implementations better be the same! This is easy to enforce by
// never defining a function in a concept map where the concept does not
// require that function.
//
// Despite the above, which relates to functions defined in different concepts
// that have a refinement relation, if a function is provided both in the
// default concept map and in a customized concept map for the same concept,
// the function in the customized map is preferred.
// [...]

In other words, it's by design that we allow the same function name in multiple concept maps, and it's required for a concept map to automatically get the functions provided in base concept maps (e.g. a concept map for a concept deriving from `Destructible` will get the `destroy` method automatically from the default concept map for `Destructible`).

**Comment by ldionne** [comment]
Based on the work in this PR, I've implemented a superset of the functionality on a local branch. I'll be merging my branch and closing this PR, since what I've implemented is a superset. Specifically, my branch checks for duplicates in concept definitions _and_ concept maps, and makes sure that we don't override a function defined in a concept that we "inherit" from. The overall compile-time penalty seems to be about 10%, which is pretty bad, but we're not really trying to make something that compiles fast right now anyway, so I'm not too worried.
